### PR TITLE
All password optional java guides

### PR DESCRIPTION
**Pull request by danielmaharry-okta**
## Description:
- **What's changed?** Add all java guides for password optional
- **Is this PR related to a Monolith release?** No
- Pulls in current changes from master branch also (by mistake)

### Resolves:

* SDK Sign up guide [OKTA-502070](https://oktainc.atlassian.net/browse/OKTA-502070)
   https://6329ad6d8d91d41a61c08187--reverent-murdock-829d24.netlify.app/docs/guides/pwd-optional-new-sign-up-email/java/main/
* SDK Sign in guide [OKTA-502078](https://oktainc.atlassian.net/browse/OKTA-502078)
   https://6329ad6d8d91d41a61c08187--reverent-murdock-829d24.netlify.app/docs/guides/pwd-optional-sign-in-email/java/main/
* Widget Sign in guide [OKTA-502089](https://oktainc.atlassian.net/browse/OKTA-502089)
  https://6329ad6d8d91d41a61c08187--reverent-murdock-829d24.netlify.app/docs/guides/pwd-optional-widget-sign-in-email/java/main/
* Change your email guide [OKTA-502084](https://oktainc.atlassian.net/browse/OKTA-502084)
   https://6329ad6d8d91d41a61c08187--reverent-murdock-829d24.netlify.app/docs/guides/pwd-optional-sign-in-email/java/main/
* Best practices guide [OKTA-503000](https://oktainc.atlassian.net/browse/OKTA-503000)
   https://6329ad6d8d91d41a61c08187--reverent-murdock-829d24.netlify.app/docs/guides/pwd-optional-best-practices/java/main/

### CODE REVIEW

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md [hunk 1]
```diff
@@ -101,45 +101,95 @@ var config = {
...
 
-![Displays the smart card IdP settings in JavaScript to remove](/img/SIW_Upgrade_Config_Change4.png)
+Remove the authentication settings for the smart card IdP (`piv`):
+
+```JavaScript
+piv: {
+  certAuthUrl: '/your/cert/validation/endpoint',
+  text: 'Authenticate with a Smart Card',
+  className: 'custom-style',
+  isCustomDomain: true,
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:166]:
```suggestion
  isCustomDomain: true
```

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md [hunk 2]
```diff
@@ -101,45 +101,95 @@ var config = {
...
 ### Bootstrapping from a recovery token
 
 If you're initializing the widget with a recovery token, the `recoveryToken` setting appears, for example:
 
-![Displays the recovery token setting](/img/SIW_Upgrade_Config_Change5.png)
+```JavaScript
+recoveryToken: 'x0whAcR02i0leKtWMZVc'
+```
 
 The recovery token is dynamic and is automatically passed into the initialization of the widget. A value in the `recoveryToken` setting currently doesn't have any effect on widget function, though, the setting takes effect in the future.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:178]:
```suggestion
The recovery token is dynamic and is automatically passed into the initialization of the widget. A value in the `recoveryToken` setting currently doesn't have any effect on widget function, though the setting takes effect in the future.
```

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md [hunk 3]
```diff
@@ -101,45 +101,95 @@ var config = {
...
 
 ### Okta dashboard or custom dashboard sign-in flow
 
-For an Okta dashboard sign-in, you no longer need to configure a redirect to the Okta Identity Cloud, create an Okta session, and then open a URL specified in the widget. You can remove the redirect configuration as follows.
+For an Okta dashboard sign-in, you no longer need to configure a redirect to the Okta Identity Cloud, create an Okta session, and then open a URL specified in the widget.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:182]:
```suggestion
For an Okta dashboard sign-in flow, you no longer need to configure a redirect to the Okta Identity Cloud, create an Okta session, and then open a URL specified in the widget.
```

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md [hunk 4]
```diff
@@ -101,45 +101,95 @@ var config = {
...
 
-![Displays the Okta dashboard sign-in](/img/SIW_Upgrade_Config_Change6.png)
+Remove the redirect configuration (`setCookieAndRedirect()`) line, shown in the following snippet:
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:184]:
```suggestion
Remove the redirect configuration (`setCookieAndRedirect()`) line shown in the following snippet:
```

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsteps.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+<div class="half border">
+
+![A form showing a list of authenticators to enroll, and Submit, Cancel, and Skip buttons.](/img/pwd-optional/pwd-optional-sign-up-authenticators-page.png)
+
+</div>
+
+### 6. The user skips the remaining optional authenticators
+
+When the user clicks the **Skip** button, call `IDXAuthenticationWrapper.skipAuthenticatorEnrollment()` method. After the user skips the remaining optional authenticators, the current status of the authentication process is now `SKIP_COMPLETE`. Call `AuthenticationResponse.getTokenResponse()` to retrieve the required access, refresh and ID tokens to pass it into your application. The user has now signed in.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsteps.md:91]:
```suggestion
When the user clicks the **Skip** button, call the `IDXAuthenticationWrapper.skipAuthenticatorEnrollment()` method. After the user skips the remaining optional authenticators, the current status of the authentication process is now `SKIP_COMPLETE`. Call `AuthenticationResponse.getTokenResponse()` to retrieve the required access, refresh, and ID tokens to pass it into your application. The user has now signed in.
```

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/java/widgetsignin.md [hunk 6]
```diff
@@ -1,3 +1 @@
-[Embedded Sign-in Widget: Sign in with email only](/docs/guides/pwd-optional-widget-sign-in-email/nodeexpress/main/)
-
-<!-- TODO: update link -->
+[Embedded Sign-in Widget: Sign in with email only](/docs/guides/pwd-optional-widget-sign-in-email/java/main/)
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/java/widgetsignin.md:1]:
```suggestion
[Embedded Sign-In Widget: Sign in with email only](/docs/guides/pwd-optional-widget-sign-in-email/java/main/)
```

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/aspnet/widgetsignin.md [hunk 7]
```diff
@@ -1,3 +1,2 @@
-[Embedded Sign-in Widget: Sign in with email only](/docs/guides/pwd-optional-widget-sign-in-email/nodeexpress/main/)
+[Embedded Sign-in Widget: Sign in with email only](/docs/guides/pwd-optional-widget-sign-in-email/dotnet/main/)
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/aspnet/widgetsignin.md:1]:
```suggestion
[Embedded Sign-In Widget: Sign in with email only](/docs/guides/pwd-optional-widget-sign-in-email/dotnet/main/)
```

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/react/widgetsignin.md [hunk 8]
```diff
@@ -1,3 +1 @@
-[Embedded Sign-in Widget: Sign in with email only](/docs/guides/pwd-optional-widget-sign-in-email/nodeexpress/main/)
-
-<!-- TODO: Do we need a react widget guide? ->
+[Embedded Sign-in Widget: Sign in with email only](/docs/guides/pwd-optional-widget-sign-in-email/react/main/)
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/react/widgetsignin.md:1]:
```suggestion
[Embedded Sign-in Widget: Sign In with email only](/docs/guides/pwd-optional-widget-sign-in-email/react/main/)
```

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
+The following code defines an endpoint that receives the `OTP` and `state` values from the magic link. If `state` is null or `IDX_CLIENT_CONTEXT` is not present in session state, the code assumes the callback has not come from the same browser and prompts the user accordingly.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md:1]:
```suggestion
The following code defines an endpoint that receives the `OTP` and `state` values from the magic link. If `state` is null or `IDX_CLIENT_CONTEXT` isn't present in session state, the code assumes the callback hasn't come from the same browser and prompts the user accordingly.
```

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+If successful, the `OTP` and `state` values are passed to the Widget with the rest of its configuration details.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md:3]:
```suggestion
If successful, the `OTP` and `state` values are passed to the widget with the rest of its configuration details.
```

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      request.getScheme() + "://" + request.getServerName() + ":" +
+      request.getServerPort() + request.getContextPath() + "/authorization-code/callback"
+   );
+   mav.addObject(ISSUER_URI, issuer);
+   session.setAttribute(CODE_VERIFIER, idxClientContext.getCodeVerifier());
+   return mav;
+}
+```
+
+> **Note**: For more information on magic links and OTP, including customizations and complete user journeys, see [Email Magic Links Overview](/docs/guides/email-magic-links-overview/main/).
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md:39]:
```suggestion
> **Note**: For more information on magic links and OTP, including customizations and complete user journeys, see [Email Magic Links overview](/docs/guides/email-magic-links-overview/main/).
```

**Review by susanharper-okta — APPROVED** [review]
couple things. approving to not hold you up

Had a few comments on one diagram:

packages/@okta/vuepress-site/.vuepress/public/img/pwd-optional/pwd-optional-java-widget-sign-in-summary.png

In the step 2 flow - the button name should be bulleted:

Click **Send me an email** button

Last line of diagram

Send user to default home page for succesful sign-in flows

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-sign-in-email/main/java/integrationsteps.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+       return verifyForm(response);
+
+   // other case statements
+
+   default:
+       return unsupportedPolicy();
+}
+```
+
+The email authenticator supports user verification by One-Time Password (OTP) and by magic links. To learn more, see the [Okta email integration guide](/docs/guides/authenticators-okta-email/java/main/#integrate-email-challenge-with-magic-links).
```
Comment (robertnicolo-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-sign-in-email/main/java/integrationsteps.md:40]: Using by is a little awkward. Consider changing to: "The email authenticator supports user verification via One-Time Password (OTP) and magic links. "

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md (hunk 9 above)
Comment (robertnicolo-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md:1]: "... prompts the user accordingly" - Where is the prompt in the sample code for this step?

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+```java
+@GetMapping("/")
+public String home(
+   @RequestParam(name = "error", required = false) String error,
+   @RequestParam(name = "interaction_code", required = false) String interactionCode,
+   @RequestParam(name = "state", required = false) String state,
+   HttpSession session) {
+
+   if (interactionCode != null && state != null) {
```
Comment (robertnicolo-okta) [inline packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md:55]: There is a prompt here but it's not called out in the step's description. What does it do?

**Review by robertnicolo-okta — CHANGES_REQUESTED** [review]
Looking good Dan. A few feedback items for you.

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-mfa-enroll-policy/main/index.md [hunk 14]
```diff
@@ -28,6 +28,8 @@ The following are the main behavior changes to the [authenticator enrollment pol
 - Existing authenticator enrollment policies (that is, policies created before the Identity Engine upgrade) still contain factors in their settings.
 - For authenticator enrollment policies modified from the Admin Console of an Identity Engine org, existing factors are converted to authenticators in their settings.
 
+> **Note:** For Identity Engine orgs with the Authenticator enrollment policy feature enabled, the new default authenticator enrollment policy created by Okta uses the authenticators setting schema. Existing default authenticator enrollment policies from a migrated org remains unchanged and still uses the factors setting schema. If you want the default policy to use the authenticators setting schema, then modify the default policy from the Admin Console to convert the settings from factors to authenticators.
```
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/oie-upgrade-mfa-enroll-policy/main/index.md:31]: fixed

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md [hunk 15]
```diff
@@ -101,45 +101,95 @@ var config = {
...
+```JavaScript
+features: {
+  idpDiscovery: true
+}
+```
 
 ### OpenID Connect/social authentication
 
-When [external Identity Providers (IdPs)](/docs/guides/identity-providers/) are used for Social Login in Classic Engine, the supported IdPs (such as Google, Facebook, Apple, Microsoft, and so on) are declared with a `type` property. These supported IdPs get distinct styling and default i18n text, while other custom entries receive general styling and require you to provide text. Each IdP can have additional CSS classes added through an optional `className` property. This feature still functions in Identity Engine. However, it's no longer the preferred method to enable the link for users to initialize the route. You can replace this method by configuring a routing rule with the application context. See [Okta Sign-In Widget with an external Identity Provider](/docs/guides/add-an-external-idp/openidconnect/main/#okta-sign-in-widget).
+You no longer require the `idps` JavaScript object in the widget, and can remove it.
```
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:143]: fixed

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md [hunk 16]
```diff
@@ -101,45 +101,95 @@ var config = {
...
+```JavaScript
+idps: [
+  { type: 'GOOGLE', id: '0oagjkh' },
+  { type: 'FACEBOOK', id: '0oagjkh' },
+    ...
+]
+```
 
-![Displays the IdPs object and className property in JavaScript to remove](/img/SIW_Upgrade_Config_Change3.png)
+This is now optional as the Sign-In Widget will automatically include IdPs based on Identity Engine routing rules.
```
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:153]: fixed

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md (hunk 2 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:178]: fixed

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md (hunk 3 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:182]: fixed

File: packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md (hunk 4 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/oie-upgrade-sign-in-widget/main/index.md:184]: fixed

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsteps.md (hunk 5 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsteps.md:91]: fixed

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/aspnet/widgetsignin.md (hunk 7 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/aspnet/widgetsignin.md:1]: fixed

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md (hunk 9 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md:1]: @susanharper-okta Gah - keep forgetting contractions.

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md (hunk 10 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md:3]: fixed

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsummary.md [hunk 17]
```diff
@@ -0,0 +1,5 @@
+<div class="full">
+
+![Sequence diagram illustrating the password-optional sign-up use case.](/img/pwd-optional/pwd-optional-java-sign-up-summary.png)
```
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsummary.md:3]: @robertnicolo-okta Have fixed point 1 but won't change for points 2 and 3.  Point 2 supports calling register() which is in the diagram and I don;t see why you need to put "check for skip" on the diagram. It's not on any of the other diagrams.

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsteps.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsteps.md:91]: good call. fixed.

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsteps.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+    idxAuthenticationWrapper.skipAuthenticatorEnrollment(proceedContext)
+
+if (response.getAuthenticationStatus() == SKIP_COMPLETE) {
+   ModelAndView modelAndView = homeHelper.proceedToHome(response.getTokenResponse(), session);
+   modelAndView.addObject("info", response.getErrors());
+   return modelAndView;
+}
+```
+
+Store these tokens for future requests and redirect the user to the default page after a successful sign-up attempt. In the example above, this is achieved in [`homeHelper`](https://github.com/okta/okta-idx-java/blob/master/samples/embedded-auth-with-sdk/src/main/java/com/okta/spring/example/helpers/HomeHelper.java)`.proceedToHome()`.
```
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/java/integrationsteps.md:104]: fixed

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/aspnet/signinwithemail.md [hunk 19]
```diff
@@ -1,3 +1 @@
-[Sign in with email only](/docs/guides/pwd-optional-sign-in-email/nodeexpress/main/)
-
-<!-- TODO: update link -->
+[Sign in with email only](/docs/guides/pwd-optional-sign-in-email/dotnet/main/)
```
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/aspnet/signinwithemail.md:1]: oops - my bad. fixed

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-sign-in-email/main/java/integrationsteps.md (hunk 12 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-sign-in-email/main/java/integrationsteps.md:40]: fixed

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md (hunk 13 above)
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md:55]: Not sure what you mean Rob. This is the code that sends the interaction code offf to exchange for an access token.

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/aspnet/bestpractices.md [hunk 20]
```diff
@@ -1,3 +1 @@
-[Best practices for password optional]()
-
-<!-- TODO: update link -->
+[Best practices for password optional](/docs/guides/pwd-optional-best-practices/aspnet/main/)
```
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-overview/main/aspnet/bestpractices.md:1]: The navbar was linking to the wrong thing. Fixed.

File: packages/@okta/vuepress-site/docs/guides/pwd-optional-sign-in-email/main/java/whatyouneed.md [hunk 21]
```diff
@@ -0,0 +1,2 @@
+* [Download and set up the SDK](/docs/guides/oie-embedded-common-download-setup-app/java/main/)
```
[the remark this answers is not in the document]
Comment (danielmaharry-okta) [reply packages/@okta/vuepress-site/docs/guides/pwd-optional-sign-in-email/main/java/whatyouneed.md:1]: fixed

**Review by robertnicolo-okta — APPROVED** [review]
lgtm

**Pull request merged**